torch/distributions.py

```diff
@@ -262,3 +262,107 @@ def log_prob(self, value):
         return (self.alpha * torch.log(self.beta) +
                 (self.alpha - 1) * torch.log(value) -
                 self.beta * value - torch.lgamma(self.alpha))
+
+
+def _dirichlet_sample_nograd(alpha):
+    gammas = torch._C._standard_gamma(alpha)
+    return gammas / gammas.sum(-1, True)
+
+
+class _Dirichlet(Function):
+    @staticmethod
+    def forward(ctx, alpha):
+        x = _dirichlet_sample_nograd(alpha)
+        ctx.save_for_backward(x, alpha)
+        return x
+
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, grad_output):
+        x, alpha = ctx.saved_tensors
+        total = alpha.sum(-1, True).expand_as(alpha)
+        grad = torch._C._dirichlet_grad(x, alpha, total)
+        return grad_output * grad
+
+
+class Dirichlet(Distribution):
+    r"""
+    Creates a Dirichlet distribution parameterized by concentration `alpha`.
+
+    Example::
+
+        >>> m = Dirichlet(torch.Tensor([0.5, 0.5]))
+        >>> m.sample()  # Dirichlet distributed with concentrarion alpha
+         0.1046
+         0.8954
+        [torch.FloatTensor of size 2]
+
+    Args:
+        alpha (Tensor or Variable): concentration parameter of the distribution
+    """
+    has_rsample = True
+
+    def __init__(self, alpha):
+        self.alpha = alpha
+
+    def _sample(self, alpha):
+        if not isinstance(alpha, Variable):
+            return _dirichlet_sample_nograd(alpha)
+        return _Dirichlet.apply(alpha)
+
+    def sample(self):
+        return self._sample(self.alpha)
+
+    def sample_n(self, n):
+        return self._sample(_expand_n(self.alpha, n))
+
+    def log_prob(self, value):
+        return ((torch.log(value) * (self.alpha - 1.0)).sum(-1) +
+                torch.lgamma(self.alpha.sum(-1)) -
+                torch.lgamma(self.alpha).sum(-1))
+
+
+class Beta(Distribution):
+    r"""
+    Creates a Beta distribution parameterized by concentration `alpha` and `beta`.
+
+    Example::
+
+        >>> m = Beta(torch.Tensor([0.5]), torch.Tensor([0.5]))
+        >>> m.sample()  # Beta distributed with concentrarion alpha
+         0.1046
+        [torch.FloatTensor of size 2]
+
+    Args:
+        alpha (Tensor or Variable): concentration parameter of the distribution
+    """
+    has_rsample = True
+
+    def __init__(self, alpha, beta):
+        alpha_num = isinstance(alpha, Number)
+        beta_num = isinstance(beta, Number)
+        if alpha_num and beta_num:
+            alpha_beta = torch.Tensor([alpha, beta])
+        else:
+            if alpha_num and not beta_num:
+                alpha = beta.new(beta.size()).fill_(alpha)
+            elif not alpha_num and beta_num:
+                beta = alpha.new(alpha.size()).fill_(beta)
+            elif alpha.size() != beta.size():
+                raise ValueError('Expected alpha.size() == beta.size(), actual {} vs {}'.format(
+                    alpha.size(), beta.size()))
```

Comment:
This seems to be a recurring pattern so it would be nice to have a helper for this. Neeraj had a neat way of dealing with it (upcasting everything to tensors, and then a tiny helper for broadcasting them together)